### PR TITLE
lightning: add retry and early terminate checking empty table

### LINKED ISSUE #31797: lightning get stuck when check table empty failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run tidb-lightning with 1k tables, there is a low probability that tidb-lightning will stuck.

From the go routine, I found it was stuck at https://github.com/pingcap/tidb/blob/7fd298d916c30ed0060db3581be58b448325819e/br/pkg/lightning/restore/check_info.go#L1132

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: close #31797

Problem Summary:

### What is changed and how it works?
- Add check `ctx.Done` when send msg to chan
- Add retry for exec `select` to avoid tidb failure. This can alleviate #31318
- Avoid output misleading log in tidb backend

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Add retry to avoid precheck failure when query execution timeout
```

### CODE REVIEW

File: br/pkg/lightning/restore/check_info.go [hunk 1]
```diff
@@ -1099,9 +1099,20 @@ func (rc *Controller) checkTableEmpty(ctx context.Context) error {
...
 	ch := make(chan string, concurrency)
 	eg, gCtx := errgroup.WithContext(ctx)
+
+	// TODO: seems there is a bug in fail-point when inject a fail-point in a function closure
+	// after fail-point fix this bug, we should move this fail-point into eg.GO
+	// See: https://github.com/pingcap/failpoint/issues/70
+	var mockErr error
+	failpoint.Inject("CheckTableEmptyFailed", func() {
+		mockErr = errors.New("mock error")
+	})
```
Comment (gozssky) [inline br/pkg/lightning/restore/check_info.go:1109]: How about adding this failpoint inside `tableContainsData`?

**Review by gozssky — APPROVED** [review]

File: br/pkg/lightning/common/util.go [hunk 2]
```diff
@@ -119,8 +119,8 @@ outside:
 		switch {
 		case err == nil:
 			return nil
-		// do not retry NotFound error
-		case errors.IsNotFound(err):
+		// do not retry NotFound error or ErrNoRows
+		case errors.IsNotFound(err), errors.ErrorEqual(err, sql.ErrNoRows):
 			break outside
 		case utils.IsRetryableError(err):
```
Comment (kennytm) [inline br/pkg/lightning/common/util.go:125]: shouldn't the `ErrNoRows` error be caught in `IsRetryableError` and thus falling into the `default:` branch?

File: br/pkg/lightning/restore/table_restore.go [hunk 3]
```diff
@@ -710,8 +710,8 @@ func (tr *TableRestore) postProcess(
 	}
 
 	// tidb backend don't need checksum & analyze
-	if !rc.backend.ShouldPostProcess() {
-		tr.logger.Debug("skip checksum & analyze, not supported by this backend")
+	if rc.cfg.PostRestore.Checksum == config.OpLevelOff && rc.cfg.PostRestore.Analyze == config.OpLevelOff {
+		tr.logger.Debug("skip checksum & analyze, either because not supported by this backend or manually disabled")
```
Comment (kennytm) [inline br/pkg/lightning/restore/table_restore.go:714]: are these (f52820108600c86757531fc3b182fb8012a00b9a) part of this PR?

Comment (glorv) [reply br/pkg/lightning/restore/table_restore.go:714]: No. As this is just a tiny change, maybe it not worth to file another PR(as while as an issue)😅。

File: br/pkg/lightning/common/util.go (hunk 2 above)
Comment (glorv) [reply br/pkg/lightning/common/util.go:125]: yes, this change is useless, I will remove it.

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**